pkg/registry/flowcontrol/rest/storage_flowcontrol.go

```diff
@@ -163,106 +147,100 @@ func shouldEnsureSuggested(flowcontrolClientSet flowcontrolclient.FlowcontrolV1b
-			}
-			continue
-		}
-		if !apierrors.IsNotFound(err) {
-			return fmt.Errorf("failed getting PriorityLevelConfiguration %s due to %v, will retry later", expectedPriorityLevelConfiguration.Name, err)
-		}
-		_, err = flowcontrolClientSet.PriorityLevelConfigurations().Create(context.TODO(), expectedPriorityLevelConfiguration, metav1.CreateOptions{FieldManager: thisFieldManager})
-		if apierrors.IsAlreadyExists(err) {
-			klog.V(3).Infof("Mandatory PriorityLevelConfiguration %s already exists, skipping creating", expectedPriorityLevelConfiguration.Name)
-			continue
-		}
-		if err != nil {
-			return fmt.Errorf("cannot create mandatory PriorityLevelConfiguration %s due to %v", expectedPriorityLevelConfiguration.Name, err)
+	for _, priorityLevelConfiguration := range priorityLevels {
+		if err := ensurePriorityLevel(clientset, objectType, priorityLevelConfiguration, plStrategy); err != nil {
+			return err
 		}
-		klog.V(3).Infof("Created mandatory PriorityLevelConfiguration %s", expectedPriorityLevelConfiguration.Name)
 	}
 	return nil
 }
 
-func flowSchemaHasWrongSpec(expected, actual *flowcontrolv1beta1.FlowSchema) (bool, error) {
+func flowSchemaSpecChanged(expected, actual *flowcontrolv1beta1.FlowSchema) bool {
 	copiedExpectedFlowSchema := expected.DeepCopy()
 	flowcontrolapisv1beta1.SetObjectDefaults_FlowSchema(copiedExpectedFlowSchema)
-	return !equality.Semantic.DeepEqual(copiedExpectedFlowSchema.Spec, actual.Spec), nil
+	return !equality.Semantic.DeepEqual(copiedExpectedFlowSchema.Spec, actual.Spec)
 }
 
-func priorityLevelHasWrongSpec(expected, actual *flowcontrolv1beta1.PriorityLevelConfiguration) (bool, error) {
+func priorityLevelSpecChanged(expected, actual *flowcontrolv1beta1.PriorityLevelConfiguration) bool {
 	copiedExpectedPriorityLevel := expected.DeepCopy()
 	flowcontrolapisv1beta1.SetObjectDefaults_PriorityLevelConfiguration(copiedExpectedPriorityLevel)
-	return !equality.Semantic.DeepEqual(copiedExpectedPriorityLevel.Spec, actual.Spec), nil
+	return !equality.Semantic.DeepEqual(copiedExpectedPriorityLevel.Spec, actual.Spec)
+}
+
+// shouldUpdateSpec inspects the auto-update annotation key and generation field to determine
+// whether the configuration object should be auto-updated.
+func shouldUpdateSpec(accessor metav1.Object) bool {
+	value, _ := accessor.GetAnnotations()[flowcontrolv1beta1.AutoUpdateAnnotationKey]
+	if autoUpdate, err := strconv.ParseBool(value); err == nil {
+		return autoUpdate
+	}
+
+	// We are here because of either a or b:
+	// a. the annotation key is missing.
+	// b. the annotation key is present but the value does not represent a boolean.
+	// In either case, if the operator hasn't changed the spec, we can safely auto update.
+	// Please note that we can't protect the changes made by the operator in the following scenario:
+	// - The operator changes the spec and then deletes and recreates the same object. (generation resets to 1)
```

Comment:
The scenario here is too specific, it excludes the case where the divergent spec appears only after the delete.